modules/distributor/forwarder/manager.go

```diff
@@ -0,0 +1,370 @@
+	defaultQueueSize   = 100
+)
+
+type Overrides interface {
+	TenantIDs() []string
+	Forwarders(tenantID string) []string
+}
+
+type Manager struct {
+	services.Service
+	logger    log.Logger
+	reg       prometheus.Registerer
+	overrides Overrides
+
+	// forwarderNameToForwarder is static throughout lifecycle of the manager and read-only
+	forwarderNameToForwarder map[string]Forwarder
+
+	tenantToQueueList   map[string]*queueList
+	tenantToQueueListMu *sync.RWMutex
+}
+
+func NewManager(cfgs ConfigList, logger log.Logger, reg prometheus.Registerer, overrides Overrides) (*Manager, error) {
+	if err := cfgs.Validate(); err != nil {
+		return nil, fmt.Errorf("failed to validate config list: %w", err)
+	}
+
+	forwarderNameToForwarder := make(map[string]Forwarder, len(cfgs))
+	for i, cfg := range cfgs {
+		forwarder, err := New(cfg, logger)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create forwarder for cfg at index=%d: %w", i, err)
+		}
+
+		forwarderNameToForwarder[cfg.Name] = forwarder
+	}
+
+	m := &Manager{
+		logger:                   logger,
+		reg:                      reg,
+		overrides:                overrides,
+		forwarderNameToForwarder: forwarderNameToForwarder,
+		tenantToQueueList:        make(map[string]*queueList),
+		tenantToQueueListMu:      &sync.RWMutex{},
+	}
+
+	m.Service = services.NewBasicService(m.start, m.run, m.stop)
+
+	return m, nil
+}
+
+func (m *Manager) ForTenant(tenantID string) List {
+	list, found := m.getQueueList(tenantID)
+	if !found {
+		return nil
+	}
+
+	return list.copy()
+}
+
+func (m *Manager) start(_ context.Context) error {
+	tenantToForwarderNames := m.readOverrides()
+	m.updateTenantToForwarderList(tenantToForwarderNames)
+
+	return nil
+}
+
+func (m *Manager) run(ctx context.Context) error {
+	ticker := time.NewTicker(1 * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			if err := m.shutdown(); err != nil {
+				return fmt.Errorf("failed to shutdown: %w", err)
+			}
+
+			return nil
+		case <-ticker.C:
+			m.handleTick()
+		}
+	}
+}
+
+func (m *Manager) stop(err error) error {
+	if err != nil {
+		_ = level.Warn(m.logger).Log("msg", "manager returned error from running state", "err", err)
+	}
+
+	return nil
+}
+
+func (m *Manager) handleTick() {
+	tenantToForwarderNames := m.readOverrides()
+	m.updateTenantToForwarderList(tenantToForwarderNames)
+}
+
+// readOverrides returns a mapping between tenant ID and a
+// list of requested forwarders.
+func (m *Manager) readOverrides() map[string][]string {
+	result := make(map[string][]string)
+
+	for _, tenantID := range m.overrides.TenantIDs() {
+		names := m.overrides.Forwarders(tenantID)
+		if len(names) < 1 {
+			continue
+		}
+
+		result[tenantID] = names
+	}
+
+	return result
+}
+
+func (m *Manager) updateTenantToForwarderList(tenantToForwarderNames map[string][]string) {
+	m.tenantToQueueListMu.Lock()
+	defer m.tenantToQueueListMu.Unlock()
+
+	oldTenantIDs := make([]string, 0, len(m.tenantToQueueList))
```

Comment:
i feel like this is unnecessarily complex. why don't we just iterate through `m.tenantToQueueList`. if the tenant is not in `tenantToForwaderNames` then call remove?

we can then call update on anything remaining.

Reply:
I don't think it's that simple. This solution does not cover new tenant IDs showing up in tenantToForwarderNames. To be honest, I like the "diff" solution. Perhaps it's a little more code than if I were to just iterate over those two maps, but IMO it's very descriptive.

Reply:
I propose this instead:

- queueList is immutable. it cannot be updated (cleans up all that code)
- on update of forwarder manager
  - loop through all existing queueLists. remove all that are to be deleted and all that need to be updated
  - call shutdown on all of these queueLists (removed and updated). this will cause them to properly drain async.
  - readd new queueLists for all updated ones that will be used by future pushes

this will clean up all the diffing code and all of the queueList update code. It will also simplify this code significantly. Since queue lists become immutable you can likely also remove all locking code there.

Reply:
I like this proposal 👍 It's very close to the logic from the original forwarder, so it makes sense to me.

Reply:
Done in [57fde51](https://github.com/grafana/tempo/pull/1775/commits/57fde5127c307c25414c4c5eb2bfe6afeeb1f818).

`queueList` is now immutable